### PR TITLE
Provide methods to train with validation context and initial predictor

### LINKED ISSUE #752: Early stopping in LightGbm
I am trying to implement a LightGbm classifier with early stopping, but I can not figure how to enter a validation set into the pipeline. I can see that EarlyStoppingRound can be set for LightGbm, so I suppose there should be a way to set a validation set. My current code:

```
var pipeline = new LearningPipeline
{
    new TextLoader(TrainDataPath).CreateFrom<MyFeature>(useHeader: true),
    new ColumnConcatenator("Features", ... ),
    new LightGbmBinaryClassifier
    {
        NumLeaves = 31, 
        NumBoostRound = 100, 
        EarlyStoppingRound = 10,
    },
};
_model = pipeline.Train<MyFeature, MyPrediction>();

```

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
Tests incoming...
Fix #752

### CODE REVIEW

File: src/Microsoft.ML.FastTree/RandomForestClassification.cs [hunk 1]
```diff
@@ -215,6 +215,8 @@ protected override Test ConstructTestForTrainingData()
         protected override BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>> MakeTransformer(IPredictorWithFeatureWeights<float> model, Schema trainSchema)
          => new BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>>(Host, model, trainSchema, FeatureColumn.Name);
 
+        public BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>> Train(IDataView trainData, IDataView validationData = null, IPredictor initialPredictor = null) => TrainTransformer(trainData, validationData, initialPredictor);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.FastTree/RandomForestClassification.cs]: >IPredictor initialPredictor  [](start = 140, length = 28)

predictor is not supported

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs [hunk 2]
```diff
@@ -275,6 +275,8 @@ protected override void InitializeTests()
         protected override BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>> MakeTransformer(IPredictorWithFeatureWeights<float> model, Schema trainSchema)
         => new BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>>(Host, model, trainSchema, FeatureColumn.Name);
 
+        public BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>> Train(IDataView trainData, IDataView validationData = null) => TrainTransformer(trainData, validationData);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.FastTree/FastTreeClassification.cs]: >TrainTransformer [](start = 143, length = 16)

@TomFinley, @Zruty FastTree support `supportTest:true` which is not exposed in TrainTransformer method, should I extend it?

**Comment by Ivanidzo4ka** [comment]
>         /// <param name="ctx">The <see cref="RegressionContext"/>.</param>

BinaryClassificationContext

---
Refers to: src/Microsoft.ML.HalLearners/HalLearnersCatalog.cs:40 in e5f0a0c. [](commit_id = e5f0a0c96e42cc5578b8ed9c1543906a519126ef, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         /// Predict a target using a linear regression model trained with the <see cref="SymSgdClassificationTrainer"/>.

 Predict a target using a linear binary classification model trained with the

---
Refers to: src/Microsoft.ML.HalLearners/HalLearnersCatalog.cs:38 in e5f0a0c. [](commit_id = e5f0a0c96e42cc5578b8ed9c1543906a519126ef, deletion_comment = False)

File: src/Microsoft.ML.HalLearners/OlsLinearRegression.cs [hunk 3]
```diff
@@ -113,6 +113,9 @@ internal OlsLinearRegressionTrainer(IHostEnvironment env, Arguments args)
         protected override RegressionPredictionTransformer<OlsLinearRegressionPredictor> MakeTransformer(OlsLinearRegressionPredictor model, Schema trainSchema)
              => new RegressionPredictionTransformer<OlsLinearRegressionPredictor>(Host, model, trainSchema, FeatureColumn.Name);
 
+        public RegressionPredictionTransformer<OlsLinearRegressionPredictor> Train(IDataView trainData, IPredictor initialPredictor = null) => TrainTransformer(trainData, initPredictor:initialPredictor);
+
+
```
Comment (Zruty0) [inline src/Microsoft.ML.HalLearners/OlsLinearRegression.cs]: *gasp

File: src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs [hunk 4]
```diff
@@ -1460,9 +1460,9 @@ internal override void Check(IHostEnvironment env)
         {
             Host.CheckNonEmpty(featureColumn, nameof(featureColumn));
             Host.CheckNonEmpty(labelColumn, nameof(labelColumn));
-            _loss = loss?? Args.LossFunction.CreateComponent(env);
+            _loss = loss ?? Args.LossFunction.CreateComponent(env);
             Loss = _loss;
-            Info = new TrainerInfo(calibration: !(_loss is LogLoss));
+            Info = new TrainerInfo(calibration: !(_loss is LogLoss), supportIncrementalTrain:true);
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/Standard/SdcaBinary.cs]: >:true [](start = 92, length = 5)

`ctrl-k-d

File: src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs [hunk 5]
```diff
@@ -403,18 +393,22 @@ public MatrixFactorizationPredictionTransformer Fit(IDataView input)
...
             {
                 model = TrainCore(ch, trainingData, validData);
             }
 
-            return new MatrixFactorizationPredictionTransformer(Host, model, input.Schema, MatrixColumnIndexName, MatrixRowIndexName);
+            return new MatrixFactorizationPredictionTransformer(Host, model, trainData.Schema, MatrixColumnIndexName, MatrixRowIndexName);
         }
 
+        /// <summary>
+        /// Train a matrix factorization model based on the input <see cref="IDataView"/> using the roles specified by XColumn and YColumn in <see cref="MatrixFactorizationTrainer"/>.
```
Comment (wschin) [inline src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs]: XColumn and YColumn no longer exist. Please use MatrixColumnIndexKind and MatrixRowIndexKind.

**Comment by wschin** [comment]
It'd be nice to have tests for training with validation for those trainers.

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs (hunk 2 above)
Comment (wschin) [reply src/Microsoft.ML.FastTree/FastTreeClassification.cs]: If you're using command line, you can print out test scores in the training phase. Not sure if C# API should be consistent with that.

File: src/Microsoft.ML.FastTree/FastTreeRanking.cs [hunk 6]
```diff
@@ -82,7 +82,7 @@ public sealed partial class FastTreeRankingTrainer
             int minDatapointsInLeaves = Defaults.MinDocumentsInLeaves,
             double learningRate = Defaults.LearningRates,
             Action<Arguments> advancedSettings = null)
-            : base(env, TrainerUtils.MakeR4ScalarLabel(labelColumn), featureColumn, weightColumn, groupIdColumn, numLeaves, numTrees, minDatapointsInLeaves, learningRate, advancedSettings)
+            : base(env, TrainerUtils.MakeU4ScalarColumn(labelColumn), featureColumn, weightColumn, groupIdColumn, numLeaves, numTrees, minDatapointsInLeaves, learningRate, advancedSettings)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.FastTree/FastTreeRanking.cs]: Not necessary with CheckLabelCompatible override

File: src/Microsoft.ML.FastTree/RandomForestClassification.cs [hunk 7]
```diff
@@ -215,12 +215,14 @@ protected override Test ConstructTestForTrainingData()
...
 
+        public BinaryPredictionTransformer<IPredictorWithFeatureWeights<float>> Train(IDataView trainData, IDataView validationData = null)
+            => TrainTransformer(trainData, validationData);
+
         protected override SchemaShape.Column[] GetOutputColumnsCore(SchemaShape inputSchema)
         {
             return new[]
             {
                 new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false, new SchemaShape(MetadataUtils.GetTrainerOutputMetadata())),
-                new SchemaShape.Column(DefaultColumnNames.Probability, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false, new SchemaShape(MetadataUtils.GetTrainerOutputMetadata(true))),
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.FastTree/RandomForestClassification.cs]: >new SchemaShape.Column(DefaultColumnNames.Probability, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false, new SchemaShape(MetadataUtils.GetTrainerOutputMetadata(true)) [](start = 16, length = 176)

By default all our scorer checks is predictor IValueMapperDist and applies proper Binder (Binary or generic one). 
For FastTree and other predictors we wrap predictor in calibrator. Which is IValueMapperDist and Binder provides Probability column.
For FastForest we don't use any calibrator. Which leads to absence of Probability column.

So I just remove expectation of Probability column.

File: src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs [hunk 8]
```diff
@@ -113,7 +113,7 @@ internal LightGbmRankingTrainer(IHostEnvironment env, LightGbmArguments args)
             double? learningRate = null,
             int numBoostRound = LightGbmArguments.Defaults.NumBoostRound,
             Action<LightGbmArguments> advancedSettings = null)
-            : base(env, LoadNameValue, TrainerUtils.MakeR4ScalarLabel(labelColumn), featureColumn, weights, groupId, numLeaves, minDataPerLeaf, learningRate, numBoostRound, advancedSettings)
+            : base(env, LoadNameValue, TrainerUtils.MakeU4ScalarColumn(labelColumn), featureColumn, weights, groupId, numLeaves, minDataPerLeaf, learningRate, numBoostRound, advancedSettings)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs]: >MakeU4ScalarColumn [](start = 52, length = 18)

Check, can I revert it back?

File: test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithValidationSet.cs [hunk 9]
```diff
@@ -30,7 +30,7 @@ public void New_TrainWithValidationSet()
             var validData = preprocess.Transform(reader.Read(GetDataPath(TestDatasets.Sentiment.testFilename)));
 
             // Train model with validation set.
-            var trainer = ml.BinaryClassification.Trainers.StochasticDualCoordinateAscent("Label","Features");
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/Scenarios/Api/Estimators/TrainWithValidationSet.cs]: >StochasticDualCoordinateAscent [](start = 59, length = 30)

SDCA don't support validation data, so I have to switch trainer.

File: test/Microsoft.ML.Tests/TrainerEstimators/FAFMEstimator.cs [hunk 10]
```diff
@@ -21,14 +21,17 @@ public void FieldAwareFactorizationMachine_Estimator()
...
+                advancedSettings: s =>
+                 {
+                     s.Shuffle = false;
+                     s.Iters = 3;
+                     s.LatentDim = 7;
+                 });
 
             TestEstimatorCore(est, data);
+            var model = est.Fit(data);
+
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/TrainerEstimators/FAFMEstimator.cs]: not needed.

File: test/Microsoft.ML.Tests/TrainerEstimators/OnlineLinearTests.cs [hunk 11]
```diff
@@ -26,14 +26,19 @@ public void OnlineLinearWorkout()
 
             var trainData = pipe.Fit(data).Transform(data).AsDynamic;
 
-            IEstimator<ITransformer> est = new OnlineGradientDescentTrainer(Env, "Label", "Features");
-            TestEstimatorCore(est, trainData);
+            var  ogdTrainer = new OnlineGradientDescentTrainer(Env, "Label", "Features");
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/TrainerEstimators/OnlineLinearTests.cs]: >   [](start = 15, length = 2)

ctrl+k+d

**Comment by Ivanidzo4ka** [comment]
>                     .Append(new KeyToValueMappingEstimator(Env, "PredictedLabel"));

test them properly!

---
Refers to: test/Microsoft.ML.Tests/TrainerEstimators/MetalinearEstimators.cs:66 in 7da3d9f. [](commit_id = 7da3d9fdb20c9354ddb03473196142522685750e, deletion_comment = False)

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs (hunk 2 above)
Comment (TomFinley) [reply src/Microsoft.ML.FastTree/FastTreeClassification.cs]: Not yet. I think we are validating this for internal scenarios.

---
In reply to: [235558027](https://github.com/dotnet/machinelearning/pull/1709#discussion_r235558027) [](ancestors = 235558027)
